### DISCUSSION TITLE
Sending messages with blocking Connection

### DISCUSSION THREAD

**Discussion by hpux735** [diwic/dbus-rs#328] [Q&A] [answered]
I'm not sure if this is really an issue, so I'll start here.  I've got the start of a library that uses the blocking::Connection version (influenced largely from the monitor.rs example).  I've got reading and decoding messages pretty far along, but when I try to send a message (signal), not only does it not appear to send, but it looks like it kills the DBus connection for the remainder of the program's lifetime.

I've been able to trace it down to https://github.com/diwic/dbus-rs/blob/master/dbus/src/channel/ffichannel.rs#L256

But, I'm not sure what it really does, and why it might be failing.  The error that's presented is `Failed to read/write data, disconnected from D-Bus?`

**Comment by diwic** [comment, marked as answer]
Hmm. One theory could be that you try to send a really weird signal on D-Bus and so the D-Bus server disconnects you because it figures you are misbehaving. That said, the client usually errors out or something and disallows you to send that message in the first place.

What message are you sending? How are you constructing that message?

**Reply by hpux735** [reply]
It turned out to be this line in the monitor.rs example:

https://gitlab.com/hpux735/dbus-callbacks/-/blob/trunk/src/main.rs#L20

When we removed it, and used the "old" scheme for receiving messages, the problem went away.

**Reply by diwic** [reply]
Side note; is a monitor really what you want to be? Usual clients are not interested in messages sent between other peers on the bus.

**Reply by hpux735** [reply]
Yah, we're using signals as an IPC bus between a collection of other devices.  So, they all listen to the same interface and are expected to get signals from them.  That part ends up working well.

**Reply by diwic** [reply]
You don't need to become a monitor to listen to signals. You should add a signal match instead. See the `match_signal` and `unity_focused_window` examples.

**Reply by hpux735** [reply]
Oh, for sure.  That's why it works without being a monitor. :)
